datastore/providers/postgres_datastore.py

```diff
@@ -0,0 +1,232 @@
+                                )
+                        metadata = json.dumps(metadata)
+
+                        row = (doc_id, chunk.id, text, embedding, metadata)
+                        copy.write_row(row)
+
+        index_statement = (
+            "CREATE INDEX ON %s USING ivfflat (embedding vector_cosine_ops)"
+            % POSTGRES_TABLENAME
+        )
+        cur.execute(index_statement)
+
+        conn.commit()
+        cur.close()
+        conn.close()
+
+        return doc_ids
+
+    async def _query(self, queries: List[QueryWithEmbedding]) -> List[QueryResult]:
+        """
+        Takes in a list of queries with embeddings and filters and
+        returns a list of query results with matching document chunks and scores.
+        """
+        # Prepare query responses and results object
+        results: List[QueryResult] = []
+
+        # Create a connection
+        conn = psycopg.connect(**self.conn_params)
+        cur = conn.cursor()
+
+        for query in queries:
+            query_results: List[DocumentChunkWithScore] = []
+            embedding = ",".join(str(v) for v in query.embedding)
+            query_statement = (
+                "SELECT doc_id, text, metadata, 1 - (%s.embedding <=> ARRAY[%s]::vector) AS score FROM %s ORDER BY score DESC LIMIT %d;"
+                % (
+                    POSTGRES_TABLENAME,
+                    embedding,
+                    POSTGRES_TABLENAME,
+                    query.top_k,
+                )
+            )
+            cur.execute(query_statement)
+            sql_query_results = cur.fetchall()
+
+            query_results: List[DocumentChunkWithScore] = []
+            for result in sql_query_results:
+                doc_id = result[0]
+                text = result[1]
+                metadata = result[2]
+                score = result[3]
+                query_result = DocumentChunkWithScore(
+                    id=doc_id, score=score, text=text, metadata=(metadata)
+                )
+                query_results.append(query_result)
+
+        results.append(QueryResult(query=query.query, results=query_results))
+
+        cur.close()
+        conn.close()
+
+        return results
+
+    async def delete(
+        self,
+        ids: List[str] | None = None,
+        filter: DocumentMetadataFilter | None = None,
+        delete_all: bool | None = None,
+    ) -> bool:
+        # Create a connection
+        conn = psycopg.connect(**self.conn_params)
+        cur = conn.cursor()
+
+        if delete_all:
+            delete_statement = "DELETE FROM {}".format(
+                sql.Identifier(POSTGRES_TABLENAME)
+            )
+            try:
+                cur.execute(delete_statement)
+            except Exception as e:
+                print(e)
+
+        if filter:
+            counter = 0
+            delete_statement = "DELETE FROM %s" % POSTGRES_TABLENAME
+
+            for key, value in filter.dict().items():
```

Comment:
This is going to be a full table scan. Can index document_id or chunk_id to make it faster. I think we should discuss why they need it this way?